File: pair_sum_k.rb
# 30/mar/19

# Given a list of numbers and a number k, return whether any two numbers
# from the list add up to k.
#
# For example, given [10, 15, 3, 7] and k of 17,
# return true since 10 + 7 is 17.
#
# Bonus: Can you do this in one pass?
class Solution
  def find(lis, k)
    @lis = lis
    @k = k
    @n = lis.length
    scan_matrix
  end

  private

  def build_result(i, j)
    { i: i, j: j, pair: [@lis[i], @lis[j]], sum: @lis[i] + @lis[j] }
  end

  def scan_matrix
    (0..@n - 2).each do |i|
      (i + 1..@n - 1).each do |j|
        res = do_check i, j
        return res if res.first
      end
    end
    [false, nil]
  end

  def do_check(i, j)
    res = build_result i, j
    puts res
    return true, res if res[:sum] == @k

    [false, nil]
  end
end

lis = [10, 15, 3, 7]
sol = Solution.new

puts "sol.find(lis, 17) => #{sol.find(lis, 17)}"
puts "sol.find(lis, 22) => #{sol.find(lis, 22)}"
puts "sol.find(lis, 100) => #{sol.find(lis, 100)}"

File: pair_sum_k.exs
#! elixir pair_sum_k.exs
# 30/mar/19

# Given a list of numbers and a number k, return whether any two numbers
# from the list add up to k.
#
# For example, given [10, 15, 3, 7] and k of 17,
# return true since 10 + 7 is 17.
#
# Bonus: Can you do this in one pass?
defmodule Solution do
  def find(lis, k) do
    %{lis: lis, k: k, n: length(lis)} |> scan_matrix
  end

  defp scan_matrix(state) do
    for i <- 0..(state[:n] - 2), j <- (i + 1)..(state[:n] - 1) do
      build_item(state, i, j)
    end
    |> Enum.find(fn x -> x[:sum] == state[:k] end)
  end

  defp build_item(state, i, j) do
    pair = Enum.map([i, j], fn x -> Enum.at(state[:lis], x) end)
    %{i: i, j: j, pair: pair, sum: Enum.sum(pair)}
  end
end

[10, 15, 3, 7]
|> Solution.find(17)
# credo:disable-for-next-line
|> IO.inspect(charlists: :as_lists)
